### ISSUE TITLE
How to parse py.test output?

### ISSUE THREAD

**Issue by emin63** [pytest-dev/pytest#2306]
What is the recommended way to parse py.test output?

At first I thought --junitxml was the way to go but it seems like that is incompatible with the xdist plugin and also not documented much.

Then I thought pytest-json looked great, but it seems to break reading a setup.cfg file.

Is there a simple way to be able to collect the output of a py.test run in a machine readable format for easy reporting?

Thanks.

**Comment by nicoddemus** [comment]
Hi,

> At first I thought --junitxml was the way to go but it seems like that is incompatible with the xdist plugin and also not documented much.

That's not true, `pytest` with `xdist` and the `junitxml` option work as expected (all our tests at work use both successfully). Can you post what error are you having, if any?

> Then I thought pytest-json looked great, but it seems to break reading a setup.cfg file.

If so it might be a bug in `pytest-json`, could you please report the problem in the appropriate channel?

> Is there a simple way to be able to collect the output of a py.test run in a machine readable format for easy reporting?

Besides `junitxml` there's also `--result-log` but that's been deprecated for some time, although we don't have a set timeline to remove it yet.

**Comment by emin63** [comment]
Thanks for the quick response.

Perhaps I was confused about junitxml. Looking at http://doc.pytest.org/en/latest/usage.html, it seems like there is a comment that it is experimental and it does not work with pytest-xdist. Maybe that was referring to record_xml_property and not junitxml?

Issue filed for pytest-json at https://github.com/mattcl/pytest-json/issues/2.

If `junitxml` is the way to go, could you point me to any tools or specifications for the output format?

Thanks again.

**Comment by nicoddemus** [comment]
> Maybe that was referring to record_xml_property and not junitxml?

Exactly. 😁 

I have updated the docs to make the distinction clearer, see #2307.

> Issue filed for pytest-json at mattcl/pytest-json#2.

Thanks, appreciate it.

> If junitxml is the way to go, could you point me to any tools or specifications for the output format?

If you execute it with a few tests the format should be self-explanatory, but you can see the official schema [here](https://github.com/windyroad/JUnit-Schema/blob/master/JUnit.xsd).

**Comment by RonnyPfannschmidt** [comment]
the basic answer is - there is absolutely no sanely parsable and complete output for pytest

the problems with xdist vs junitxml probably relate to the fact that transferred trackebacks are basically "broken" and lead to different junitxml output

**Comment by emin63** [comment]
Just an update for anyone coming here to parse pytest output:

  1. pytest-json does not parse errors in trying to load tests/modules (see https://github.com/mattcl/pytest-json/issues/3)
  2. I've had the best luck using junitxml with the `xmltodict` module. If you really want json, you could just dump the result to json using `json.dumps`.

Thanks.

**Comment by kaczmarj** [comment]
i am wondering how to parse pytest output as well. https://github.com/pytest-dev/pytest-reportlog seems to be the right tool for me.

**Comment by RonnyPfannschmidt** [comment]
exactly, as intended - if youwant parsable output, make sure you get an artifact for that

**Comment by RonnyPfannschmidt** [comment]
that being said, it might be interesting to have normal ux on stderr and reportlog on stdout

**Comment by nimdanor** [comment]
yes it would be a good thing to be able to output a machine readable format by using a simple cli option. Like 
python -m pytest -h whatever

**Comment by jeffwright13** [comment]
> yes it would be a good thing to be able to output a machine readable format by using a simple cli option. Like python -m pytest -h whatever

I wrote [a plugin](https://github.com/jeffwright13/pytest-recap) a while back that addresses some shortcomings I saw in the existing solutions. It can produce JSON output as per [the specified format](https://github.com/jeffwright13/pytest-recap?tab=readme-ov-file#recap-json-format). The command-line to do so would be 
```
pytest --recap
```

There is also a [discussion](https://github.com/jeffwright13/pytest-recap?tab=readme-ov-file#comparison-with-other-pytest-reporting-plugins) on that repo's README that talks about the various options (that I knew of at the time) for outputting machine-readable session files.
